src/mono/wasm/runtime/invoke-js.ts

```diff
@@ -0,0 +1,316 @@
+            const covertor = cs_to_js_map.get(mono_type);
+            assert(covertor, `Unknow covertor for type ${mono_type} at ${position}`);
+            const convertor_name = `covertor${position}`;
+            closure[convertor_name] = covertor;
+
+            body += `  const arg${position}_js = closure.${convertor_name}(<any>data + ${offset});\n`;
+        }
+        if (arg_count === 0) {
+            body += "  const res_js = closure.fn();\n";
+        }
+        else if (arg_count === 1) {
+            body += "  const res_js = closure.fn(arg1_js);\n";
+        }
+
+        const res_mono_type = <MonoType><any>getU32(<any>signature);
+        if (res_mono_type === MonoTypeNull) {
+            body += "  assert(value === undefined, 'Function returned unexpected value, C# signature is void');\n";
+        } else {
+            const covertor = js_to_cs_map.get(res_mono_type);
+            assert(covertor, `Unknow covertor for type ${res_mono_type} at ret`);
+            closure["covertor0"] = covertor;
+
+            body += "  closure.covertor0(<any>data);\n";
+        }
+
+        body += "  return InvokeJSResult.Success;\n";
+        body += "} catch (ex) {\n";
+        body += "  _marshal_exception_to_cs(data, ex);\n";
+        body += "  return InvokeJSResult.FunctionThrewException;\n";
+        body += "}\n";
+
+        const bound_fn = new Function("closure", "data", body).bind(null, closure);
+        const bound_function_handle = mono_wasm_get_js_handle(bound_fn)!;
+        setI32(function_js_handle, <any>bound_function_handle);
+        return MonoObjectNull;
+    } catch (ex) {
+        return wrap_error(is_exception, ex);
+    }
+}
+
+export function mono_wasm_invoke_bound_function(function_js_handle: JSHandle, data: VoidPtr): InvokeJSResult {
+    const js_fn = mono_wasm_get_jsobj_from_js_handle(function_js_handle);
+    assert(js_fn, `Unknow function for handle ${function_js_handle}`);
+    return js_fn(data);
+}
+
+
+
+function mono_wasm_lookup_function(function_id: MonoString): Function {
+    let res = _invoke_js_function_cache.get(function_id);
+    if (res !== undefined) {
+        return res;
+    }
+
+    const js_function_name = conv_string(function_id);
+    assert(js_function_name, "js_function_name must be string");
+    res = _walk_global_scope_to_find_function(js_function_name);
+    _invoke_js_function_cache.set(function_id, res);
+
+    return res;
+}
+
+function covert_arg_to_js(signature: VoidPtr, data: VoidPtr, position: number): any {
+    //console.log("covert_arg_to_js A " + (<any>signature + (position << 2)));
+    const mono_type = <MonoType><any>getU32(<any>signature + (position << 2));
+    //console.log("covert_arg_to_js B " + mono_type);
+    const covertor = cs_to_js_map.get(mono_type);
+    //console.log("covert_arg_to_js C " + typeof covertor);
+    assert(covertor, `Unknow covertor for type ${mono_type} at ${position}`);
```

Comment:
I think this is a scenario that needs good error handling and messaging - for example we need to convert that MonoType into a string FullName because it's quite likely an end user will pass an unsupported type to JS and we can't expect them to make sense of a raw typehandle.

Reply:
Also it just occurred to me that since this 'assert' statement contains string expansions, we will probably be formatting a string every time we check the condition. Asserts like this in hot paths need to be if statements or string literals that don't have an expansion.

Reply:
Fixed both, thank you